klient/k8s/resources/decoder.go

```diff
@@ -0,0 +1,354 @@
+	"sigs.k8s.io/e2e-framework/klient/k8s"
+)
+
+func isDirectory(dir string) bool {
+	d, err := os.Stat(dir)
+	if err != nil {
+		return false
+	}
+	return d.IsDir()
+}
+
+func listResourceFiles(directory string) ([]string, error) {
+	// determine if directory is a path to a directory, or if it has a filename Glob included
+	var base string
+	dir, err := os.Stat(directory)
+	if err != nil { // if this is not a resolvable path, assume the "base" of the path is a filename-prefix
+		base = filepath.Base(directory)
+		directory = filepath.Dir(directory)
+		if !isDirectory(directory) {
+			return nil, fmt.Errorf("resources: files with prefix %q in directory %q could not be found", base, directory)
+		}
+	} else if err == nil && dir.IsDir() {
+		// it is a directory, continue
+	} else if !dir.IsDir() { // this is not a directory, but an (unexpected) file
+		return nil, fmt.Errorf("resources: %q is not a directory. not supported", dir.Name())
+	}
+	yamls, err := filepath.Glob(filepath.Join(directory, fmt.Sprintf("%s*.yaml", base)))
+	if err != nil {
+		return nil, err
+	}
+	ymls, err := filepath.Glob(filepath.Join(directory, fmt.Sprintf("%s*.yml", base)))
+	if err != nil {
+		return nil, err
+	}
+	jsons, err := filepath.Glob(filepath.Join(directory, fmt.Sprintf("%s*.json", base)))
+	if err != nil {
+		return nil, err
+	}
+	filenames := yamls
+	filenames = append(filenames, ymls...)
+	filenames = append(filenames, jsons...)
+	sort.Strings(filenames) // sort before returning files for a consistent order
+	return filenames, nil
+}
+
+// MutateFunc is a function executed after an object is decoded to alter its state in a pre-defined way, and can be used to apply defaults.
+// Returning an error halts decoding of any further objects.
+type MutateFunc func(obj k8s.Object) error
+
+// HandlerFunc is a function executed after an object has been decoded and patched. If an error is returned, further decoding is halted.
+type HandlerFunc func(ctx context.Context, obj k8s.Object) error
+
+// DecodeEachFile resolves files at the filepath (with Glob support), decoding files that match the common JSON or YAML file extensions (json, yaml, yml). Supports multi-document files.
+//
+// Example filepath:
+// `path/to/dir` -- matches appropriate files in the directory
+// `path/to/dir/file-prefix-` -- (if not also a directory) matches files in the directory that start "file-prefix-"
+//
+// If provided, the defaults GroupVersionKind is used to determine the k8s.Object underlying type, otherwise rely the innate typing of the scheme.
+// Falls back to the unstructured.Unstructured type if a matching type cannot be found for the Kind.
+//
+// If handlerFn returns an error, decoding is halted.
+// Patches are optional and applied after decoding and before handlerFn is executed.
+//
+func DecodeEachFile(ctx context.Context, directoryGlob string, defaults *schema.GroupVersionKind, handlerFn HandlerFunc, patches ...MutateFunc) error {
```

Comment:
Consider using FS from the  io/fs package (instead of taking a directoryGlob string). That way the resource yaml/json files can be from the local filesystem or even be embedded in the binary.

Reply:
Matter of fact, I think once you explore os/fs, it may make func `listResourceFiles` unnecessary.